File: Antenas/Classes/Lista.cs
namespace Antenas.Classes;

/// <summary>
/// Classe de lista de antenas
/// </summary>
public class Lista
{
    private LinkedList<Antena> lista = new();
    private LinkedList<(int, int)> listaErros = new();

    /// <summary>
    /// Adicionar uma nova antena
    /// </summary>
    /// <param name="antena">Nova Antena</param>
    public void Adicionar(Antena antena)
    {
        lista.AddLast(antena);
    }

    /// <summary>
    /// Remover uma antena
    /// </summary>
    /// <param name="antena">Antena a remover</param>
    public void Remover(int x, int y)
    {
        int count = 0;
        if (lista.Any())
        {
            var node = lista.First;
            
            while (node != null)
            {
                if (node.Value.PosX == x && node.Value.PosY == y)
                {
                    lista.Remove(node);
                    count++;
                    break;
                }
                node = node.Next;
            }

            if (count == 0)
            {
                Console.WriteLine("Nenhuma antena para remover!");
            }
            else
            {
                Console.WriteLine("Antena removida!");
            }
        }
        else
        {
            Console.WriteLine("Nenhuma antena para remover!");
        }
    }

    /// <summary>
    /// Mostrar Antenas
    /// </summary>
    public void Mostrar()
    {
        if (lista.Any())
        {
            foreach (var antena in lista)
            {
                Console.WriteLine($"Frequência: {antena.Frequencia}, Posição: ({antena.PosX}, {antena.PosY})");
            }
        }
        else
        {
            Console.WriteLine("Nenhuma antena adicionada!");
        }
    }
    
    /// <summary>
    /// Carregar ficheiro de Antenas
    /// </summary>
    /// <param name="caminho">Caminho do ficheiro</param>
    public void CarregarFicheiro(string caminho)
    {
        if (!File.Exists(caminho))
        {
            Console.WriteLine("Ficheiro não encontrado.");
            return;
        }
        
        string[] linhas = File.ReadAllLines(caminho);
        for (int i = 0; i < linhas.Length; i++)
        {
            for (int j = 0; j < linhas[i].Length; j++)
            {
                if (char.IsLetterOrDigit(linhas[i][j]))
                {
                    Antena novaAntena = new Antena(linhas[i][j], j, i);
                    Adicionar(novaAntena);
                }
            }
        }
        
        ValidarErros();
    }
    
    /// <summary>
    /// Validar Erros de Efeitos Nefastos
    /// </summary>
    public void ValidarErros()
    {
        if (lista.Any())
        {
            listaErros.Clear();
            var antenasArray = lista.ToList();

            // Calcular limites máximos com base nas antenas existentes
            int largura = antenasArray.Any() ? antenasArray.Max(a => a.PosX) + 1 : 0;
            int altura = antenasArray.Any() ? antenasArray.Max(a => a.PosY) + 1 : 0;

            for (int i = 0; i < antenasArray.Count; i++)
            {
                for (int j = i + 1; j < antenasArray.Count; j++)
                {
                    var a1 = antenasArray[i];
                    var a2 = antenasArray[j];

                    if (a1.Frequencia != a2.Frequencia) continue;

                    int dx = a2.PosX - a1.PosX;
                    int dy = a2.PosY - a1.PosY;

                    if (dx % 2 != 0 || dy % 2 != 0) continue;

                    int px1 = a1.PosX - dx;
                    int py1 = a1.PosY - dy;
                    if (Limites(px1, py1, largura, altura))
                        listaErros.AddLast((px1, py1));

                    int px2 = a2.PosX + dx;
                    int py2 = a2.PosY + dy;
                    if (Limites(px2, py2, largura, altura))
                        listaErros.AddLast((px2, py2));
                }
            }
        }
        else
        {
            Console.WriteLine("Nenhuma antena adicionada!");
        }
    }
    
    /// <summary>
    /// Calcular limites da listagem
    /// </summary>
    /// <param name="x">X</param>
    /// <param name="y">Y</param>
    /// <param name="largura">Largura da lista</param>
    /// <param name="altura">Altura da lista</param>
    /// <returns></returns>
    private bool Limites(int x, int y, int largura, int altura)
    {
        return x >= 0 && y >= 0 && x < largura && y < altura;
    }
    
    /// <summary>
    /// Mostrar listagem de erros nefastos
    /// </summary>
    public void MostrarErros()
    {
        if (listaErros.Any())
        {
            Console.WriteLine("Localizações com efeito nefasto:");
            foreach (var localizacao in listaErros)
            {
                Console.WriteLine($"Coordenadas: ({localizacao.Item1}, {localizacao.Item2})");
            }
        }
        else
        {
            Console.WriteLine("Nenhuma antena adicionada!");
        }
    }
}

File: lista.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "lista.h"

#define MAX_LINHA 256

Antena* lista = NULL;
Localizacao* efeitos = NULL;

void AdicionarAntena(char freq, int x, int y) {
    Antena* nova = (Antena*)malloc(sizeof(Antena));
    nova->frequencia = freq;
    nova->x = x;
    nova->y = y;
    nova->next = lista;
    lista = nova;
}

void RemoverAntena(int x, int y) {
    Antena **indirect = &lista;
    while (*indirect) {
        if ((*indirect)->x == x && (*indirect)->y == y) {
            Antena* temp = *indirect;
            *indirect = temp->next;
            free(temp);
            printf("Antena removida!\n");
            return;
        }
        indirect = &(*indirect)->next;
    }
    printf("Nenhuma antena para remover!\n");
}

void MostrarAntenas() {
    Antena* atual = lista;
    if (!atual) {
        printf("Nenhuma antena adicionada!\n");
        return;
    }
    while (atual) {
        printf("Frequência: %c, Posição: (%d, %d)\n", atual->frequencia, atual->x, atual->y);
        atual = atual->next;
    }
}

static int DentroDosLimites(int x, int y, int largura, int altura) {
    return x >= 0 && y >= 0 && x < largura && y < altura;
}

static void AdicionarEfeito(int x, int y) {
    Localizacao* nova = (Localizacao*)malloc(sizeof(Localizacao));
    nova->x = x;
    nova->y = y;
    nova->next = efeitos;
    efeitos = nova;
}

static void LimparEfeitos() {
    while (efeitos) {
        Localizacao* temp = efeitos;
        efeitos = efeitos->next;
        free(temp);
    }
}

void ValidarEfeitos() {
    LimparEfeitos();
    for (Antena* a1 = lista; a1; a1 = a1->next) {
        for (Antena* a2 = a1->next; a2; a2 = a2->next) {
            if (a1->frequencia != a2->frequencia) continue;
            int dx = a2->x - a1->x;
            int dy = a2->y - a1->y;
            if (dx % 2 != 0 || dy % 2 != 0) continue;
            int largura = (a1->x > a2->x ? a1->x : a2->x) + 10;
            int altura = (a1->y > a2->y ? a1->y : a2->y) + 10;
            int px1 = a1->x - dx;
            int py1 = a1->y - dy;
            int px2 = a2->x + dx;
            int py2 = a2->y + dy;
            if (DentroDosLimites(px1, py1, largura, altura)) AdicionarEfeito(px1, py1);
            if (DentroDosLimites(px2, py2, largura, altura)) AdicionarEfeito(px2, py2);
        }
    }
}

void MostrarEfeitos() {
    if (!efeitos) {
        printf("Nenhuma antena adicionada!\n");
        return;
    }
    printf("Localizações com efeito nefasto:\n");
    for (Localizacao* atual = efeitos; atual; atual = atual->next) {
        printf("Coordenadas: (%d, %d)\n", atual->x, atual->y);
    }
}

void CarregarFicheiro(const char* caminho) {
    FILE* file = fopen(caminho, "r");
    if (!file) {
        printf("Ficheiro não encontrado.\n");
        return;
    }
    char linha[MAX_LINHA];
    int y = 0;
    while (fgets(linha, sizeof(linha), file)) {
        for (int x = 0; linha[x] != '\0'; x++) {
            if (isalnum(linha[x])) {
                AdicionarAntena(linha[x], x, y);
            }
        }
        y++;
    }
    fclose(file);
    ValidarEfeitos();
}
